ndk/src/surface_texture.rs

```diff
@@ -0,0 +1,112 @@
+//! Bindings for [`ffi::ASurfaceTexture`]
+#[cfg(feature = "api-level-28")]
+use crate::native_window::NativeWindow;
+#[cfg(feature = "api-level-28")]
+use jni_sys::{jobject, JNIEnv};
+use std::ptr::NonNull;
+use thiserror::Error;
+
+#[derive(Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct SurfaceTexture {
+    ptr: NonNull<ffi::ASurfaceTexture>,
+}
+
+unsafe impl Send for SurfaceTexture {}
+
+#[derive(Debug, Error)]
+pub struct PosixError(pub i32);
+
+impl std::fmt::Display for PosixError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "Posix Error: {}", self.0)
+    }
+}
+
+impl Drop for SurfaceTexture {
+    fn drop(&mut self) {
+        unsafe { ffi::ASurfaceTexture_release(self.ptr.as_ptr()) }
+    }
+}
+
+impl SurfaceTexture {
+    /// Assumes ownership of `ptr`
+    ///
+    /// # Safety
+    /// `ptr` must be a valid pointer to an Android [`ffi::ASurfaceTexture`].
+    pub unsafe fn from_ptr(ptr: NonNull<ffi::ASurfaceTexture>) -> Self {
+        Self { ptr }
+    }
+
+    #[cfg(feature = "api-level-28")]
+    /// # Safety
+    ///
+    /// This function should be called with a healthy JVM pointer and with a non-null surface texture,
+    /// which must be kept alive on the Java/Kotlin side.
+    pub unsafe fn from_surface_texture(env: *mut JNIEnv, surface_texture: jobject) -> Option<Self> {
+        let a_surface_texture_ptr = ffi::ASurfaceTexture_fromSurfaceTexture(env, surface_texture);
+        let s = NonNull::new(a_surface_texture_ptr)?;
+        Some(SurfaceTexture::from_ptr(s))
+    }
+
+    /// Returns native internal pointer
+    pub fn ptr(&self) -> NonNull<ffi::ASurfaceTexture> {
+        self.ptr
+    }
+
+    #[cfg(feature = "api-level-28")]
+    /// Returns a reference to an ANativeWindow (i.e. the Producer)
+    /// for this SurfaceTexture. This is equivalent to Java's:
+    /// Surface sur = new Surface(surfaceTexture);
```

Comment:
How about newlining this and adding some more intradoc-links, so that the end-result looks like (feel free to apply this suggestion):


```suggestion
    /// Returns a reference to a [`NativeWindow`] (i.e. the Producer) for this [`SurfaceTexture`].
    ///
    /// This is equivalent to Java's:
    /// ```java
    /// Surface sur = new Surface(surfaceTexture);
    /// ```
```